Add Lua Scripting support to nri-redis

omnibs opened the pull request:
Add an API to do Lua scripting to `Nri.Redis`.

We have a new Handler function `eval`, which does the recommended dance:
- Calculate the SHA1 of the Lua script we intend to run
- Try to run it with `EVALSHA ${sha1}`
  - On failure:
    - `LOAD SCRIPT ${theScript}`
    - Retry `EVALSHA ${sha1}`

The failure case only happens:
- The very first time a Lua script is tried on a specific Redis server
- The very first time any changes to the Lua script are tried on a specific Redis server

I initially tried to do this as a `Redis.Api` / `Internal.Query`, but it didn't work. If I understand semantics of type classes correctly, it won't unless `Internal.Query` is a `Monad`, not an `Applicative`.

Hedis uses a different `RedisResponse` typeclass for scripting results, which has its own decoding thingy. I wrote a few orphaned instances for common data types, like `Text`, `Int` and `()`.

Notes:
- The `eval` API supports namespacing of keys
- The `eval` API **does not** support auto-extending key expiry

omnibs committed:
Write the shell of script execution

The core will be in Redis.Script

omnibs committed:
format

omnibs committed:
Attempt at an attoparsec parser, which I gave up on

Turns out error messages are important

omnibs committed:
Ditch attoparsec for megaparsec: great error msgs

omnibs committed:
Create TemplateHaskell API for writing Lua Scripts

omnibs committed:
format

omnibs committed:
Write a failing test for eval

omnibs committed:
Update srcLoc lines for tests

omnibs committed:
Implement mapKeys

omnibs committed:
Implement keysTouchedByScript

omnibs committed:
Implement evalString (test is green now)

omnibs committed:
Add failing test for script with arguments

This is failing bc I misunderstood `eval` in Redis

Eval is: EVAL script numkeys key key key arg arg arg

Keys here is not the NAME of the argument, as I thought, but rather an
input parameter in full

There's no named arguments in Redis. I assumed that by reading examples
from the C# API Client in the Rate Limiting implementation docs.

Args and Keys are referenced like KEYS[1] KEYS[2] KEYS [3] and ARGV[1]
ARGV[2] ARGV[3]

Let's fix that in the next few commits

omnibs committed:
Fix API and hence fix script w/ arguments test

Splits up keys and arguments, uses KEYS[1] and ARGV[1] inside luaScript

omnibs committed:
Add returning argument as Text

This required a RedisResult instance for Text

omnibs committed:
Warn about returning keys in scripts

omnibs committed:
Reorganize code in Script

omnibs committed:
Failed attempt at implementing a caching eval API

Redis provides a faster API for running scripts, where you:
- SCRIPT LOAD "the script"
- EVALSHA "script sha1" n_keys [keys] [args]

I can't implement this because we'd need to restrict `doRawQuery`.

`doRawQuery` is generic for both instances of `RedisCtx` in Hedis:
- Transactions: `RedisCtx RedisTx Queued`
- Regular queries: `RedisCtx Redis (Either Reply)`

I'll need a different approach for this.

omnibs committed:
Remove failed attempt at evalCached

omnibs committed:
Move eval to a handler function

Allows us to do the `evalsha -> script load -> evalsha` flow

omnibs committed:
Add orphaned instances for Int and ()

omnibs committed:
Test that lists work

omnibs committed:
Fix for GHC 8.10

omnibs committed:
Update golden files for ghc 8.10

omnibs committed:
Fix for NoRedInk monorepo

omnibs committed:
Remove Eval Cmd

The Handler eval is better. The only thing it doesn't support is
auto-expire, but I'm skeptical it's a good idea to use that with Lua
scripts arbitrarily.

omnibs committed:
ormolu

omnibs committed:
Reorg for hackage docs

micahhahn commented on nri-redis/test/Spec/Redis/Script.hs:
```diff
@@ -0,0 +1,129 @@
+        -- the lines below:
+        -- Test.test "compilation error" <| \_ ->
+        --   [script|${123}|]
+        --     |> Expect.equal "Doesn't matter, this won't compile"
```
Sooo, we can in fact test for compile-time errors :) 

The basic idea is that you call your function normally (not in a splice), and then there are tools in the `Language.Haskell.TH.TestUtils` module which let you run the `Q` monad and test the result. 

You might want to take a look at what I did in `nri-postgresql/test/Enum.hs` way back when I added postgres enum support.

micahhahn replied:
Also, let's pair if this is something you would want to add!

micahhahn commented on nri-redis/test/Spec/Redis/Script.hs:
```diff
@@ -0,0 +1,129 @@
+
+mapLeft :: (a -> c) -> Either a b -> Either c b
+mapLeft f (Left a) = Left (f a)
+mapLeft _ (Right b) = Right b
```
🎨 You could use [`mapFirst`](https://hackage.haskell.org/package/base-4.20.0.0/docs/Data-Bifunctor.html#v:first) instead from `Data.BIfunctor`

micahhahn committed:
Let's avoid the `IncoherentInstances` extension as its dangerous and deprecated

micahhahn commented on nri-redis/src/Redis/Script.hs:
```diff
@@ -0,0 +1,286 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE IncoherentInstances #-}
```
I found a nicer way to avoid this: 

- #111

micahhahn reviewed (changes requested):
Amazing work! I really appreciate us going the extra mile here to make a good interface for this. 

A couple things to change below, but nothing deal breaking.

jali-clarke committed:
Merge branch 'trunk' into phx-1356-add-eval-api-to-nri-redis

omnibs committed:
Merge pull request #111 from NoRedInk/phx-1356-add-eval-remove-global-incoherentinstances

Patch to avoid using `IncoherentInstances`

omnibs committed:
Add th-test-utils to the haskell pkgset (+format)

I ended up not needing this, but I noticed it was missing and failing
tests for nri-postgresql locally

omnibs committed:
Add type-check test for script w/o Key or Literal

omnibs committed:
Explain what this does

omnibs committed:
Use Bifunctor.first

omnibs committed:
Fix golden-results for nri-postgresql

omnibs committed:
Fix discrepancy in LoC between ghc 9 and ghc 8

omnibs committed:
Unblock local run of ObservabilitySpec.hs

In CI I think the postgres user gets created somehow. Locally it doesn't
